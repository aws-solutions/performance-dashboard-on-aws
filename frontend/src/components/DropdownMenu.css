[data-reach-menu-item] {
  cursor: pointer;
  margin-bottom: 4px;
  margin-right: 10px;
  margin-left: 10px;
  border-radius: 0.2rem;
}

[data-reach-menu-list],
[data-reach-menu-items] {
  padding: 0.7rem 0 0.8rem 0 !important;
}

[data-reach-menu-item][data-selected] {
  background: #565c65 !important;
}

[data-reach-menu-link] {
  text-decoration: none;
  color: black;
  display: block;
}

[data-reach-menu-list] > *:last-child {
  margin-bottom: 0px;
  min-width: 120px;
}

[data-reach-menu-item][aria-disabled] {
  cursor: default !important;
}

[data-reach-menu-list] {
  background-color: white;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.1);
  padding: 15px;
}

[data-reach-menu-list][tabindex]:focus {
  outline: none;
}

[data-reach-popover][data-reach-menu][data-reach-menu-popover] {
  z-index: 999999;
}
