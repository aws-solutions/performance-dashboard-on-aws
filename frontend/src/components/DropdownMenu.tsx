/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import "./DropdownMenu.css";
import { Menu, MenuList, MenuButton, MenuItem, MenuLink } from "@reach/menu-button";
import "@reach/menu-button/styles.css";
import React, { ReactNode } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCaretDown, IconDefinition } from "@fortawesome/free-solid-svg-icons";

type Variant = "base" | "default" | "outline" | "secondary" | "accent" | "unstyled";

interface DropdownProps {
    children: ReactNode;
    variant?: Variant;
    className?: string;
    ariaLabel?: string;
    buttonText: string;
    disabled?: boolean;
    icon?: IconDefinition;
}

function DropdownMenu(props: DropdownProps) {
    let variantClassName = "";
    switch (props.variant) {
        case "base":
            variantClassName = " usa-button--base";
            break;
        case "outline":
            variantClassName = " usa-button--outline";
            break;
        case "secondary":
            variantClassName = " usa-button--secondary";
            break;
        case "accent":
            variantClassName = " usa-button--accent-cool";
            break;
        case "unstyled":
            variantClassName = " usa-button--unstyled";
            break;
        default:
            variantClassName = " usa-button--base";
            break;
    }

    let additionalClasses = "";
    if (props.className) {
        additionalClasses = ` ${props.className}`;
    }

    return (
        <Menu>
            <MenuButton
                className={`usa-button${variantClassName}${additionalClasses}`}
                aria-label={props.ariaLabel}
                disabled={props.disabled}
            >
                {props.buttonText}
                <FontAwesomeIcon icon={props.icon ?? faCaretDown} className="margin-left-1" />
            </MenuButton>
            <MenuList className="font-sans-xs">{props.children}</MenuList>
        </Menu>
    );
}

DropdownMenu.MenuItem = MenuItem;
DropdownMenu.MenuLink = MenuLink;

export default DropdownMenu;
