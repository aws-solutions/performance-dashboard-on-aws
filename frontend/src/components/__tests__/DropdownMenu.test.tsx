/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render } from "@testing-library/react";
import DropdownMenu from "../DropdownMenu";
import { MenuItem } from "@reach/menu-button";

test("renders a default dropdown menu", async () => {
    const wrapper = render(
        <DropdownMenu buttonText="Example text">
            <MenuItem onSelect={() => {}}> This is a menu item</MenuItem>
        </DropdownMenu>,
    );
    expect(wrapper.container).toMatchSnapshot();
});

test("renders an outlined menu button", async () => {
    const wrapper = render(
        <DropdownMenu buttonText="Example text" variant="outline">
            <MenuItem onSelect={() => {}}> This is a menu item</MenuItem>
        </DropdownMenu>,
    );
    expect(wrapper.container).toMatchSnapshot();
});

test("renders a secondary menu button", async () => {
    const wrapper = render(
        <DropdownMenu buttonText="Example text" variant="secondary">
            <MenuItem onSelect={() => {}}> This is a menu item</MenuItem>
        </DropdownMenu>,
    );
    expect(wrapper.container).toMatchSnapshot();
});

test("renders an accent menu button", async () => {
    const wrapper = render(
        <DropdownMenu buttonText="Example text" variant="accent">
            <MenuItem onSelect={() => {}}> This is a menu item</MenuItem>
        </DropdownMenu>,
    );
    expect(wrapper.container).toMatchSnapshot();
});

test("renders an unstyled menu button", async () => {
    const wrapper = render(
        <DropdownMenu buttonText="Example text" variant="unstyled">
            <MenuItem onSelect={() => {}}> This is a menu item</MenuItem>
        </DropdownMenu>,
    );
    expect(wrapper.container).toMatchSnapshot();
});
