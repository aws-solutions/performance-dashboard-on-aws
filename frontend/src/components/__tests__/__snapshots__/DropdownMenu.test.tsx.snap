// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a default dropdown menu 1`] = `
<div>
  <button
    aria-controls="menu--1"
    aria-haspopup="true"
    class="usa-button usa-button--base"
    data-reach-menu-button=""
    id="menu-button--menu--1"
    type="button"
  >
    Example text
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-caret-down fa-w-10 margin-left-1"
      data-icon="caret-down"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 320 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`renders a secondary menu button 1`] = `
<div>
  <button
    aria-controls="menu--12"
    aria-haspopup="true"
    class="usa-button usa-button--secondary"
    data-reach-menu-button=""
    id="menu-button--menu--12"
    type="button"
  >
    Example text
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-caret-down fa-w-10 margin-left-1"
      data-icon="caret-down"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 320 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`renders an accent menu button 1`] = `
<div>
  <button
    aria-controls="menu--17"
    aria-haspopup="true"
    class="usa-button usa-button--accent-cool"
    data-reach-menu-button=""
    id="menu-button--menu--17"
    type="button"
  >
    Example text
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-caret-down fa-w-10 margin-left-1"
      data-icon="caret-down"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 320 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`renders an outlined menu button 1`] = `
<div>
  <button
    aria-controls="menu--7"
    aria-haspopup="true"
    class="usa-button usa-button--outline"
    data-reach-menu-button=""
    id="menu-button--menu--7"
    type="button"
  >
    Example text
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-caret-down fa-w-10 margin-left-1"
      data-icon="caret-down"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 320 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`renders an unstyled menu button 1`] = `
<div>
  <button
    aria-controls="menu--22"
    aria-haspopup="true"
    class="usa-button usa-button--unstyled"
    data-reach-menu-button=""
    id="menu-button--menu--22"
    type="button"
  >
    Example text
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-caret-down fa-w-10 margin-left-1"
      data-icon="caret-down"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 320 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;
